File: seeker/snippet/stack.py
#date: 2022-08-25T16:53:18Z
#url: https://api.github.com/gists/4540fb821d21adee5df29609aec131bb
#owner: https://api.github.com/users/jermspeaks

# A "first-in-last-out" data structure
class StackEmptyError(IndexError):
    """Attempt to pop an empty stack."""

class Stack:
    def __init__(self):
        self._list = []
  
    def pop():
        # Removes the top item from the stack
        # should we check if list is not empty? => yep
        if not self._list:
            raise StackEmptyError()
        else:
            return self._list.pop()
    
        


    def push(item):
        # Adds an item to the top of the stack
        self._list.append(item)


    def peek():
        # Return the item at the front of the queue (but do not remove it)
        if not self._list:
            return StackEmptyError
        else:
            return self._list[-1]

    # def isEmpty():
    #     # Check whether the queue is empty
    

  # get length() {
  #   // Return the length of the stack

# let books = new Stack()
# books.push("Book 1")
# books.push("Book 2")
# books.push("Book 3")
# print books.peek())
# books.pop()
# print books.peek()
# print books.length

File: seeker/snippet/stack.go
//date: 2023-11-02T16:33:56Z
//url: https://api.github.com/gists/be12e78061eddcf69b89d56587641a87
//owner: https://api.github.com/users/mmtdebojit

package iterator

type Stack struct {
	items []interface{}
}

func (s *Stack) Push(item interface{}) {
	s.items = append(s.items, item)
}

func (s *Stack) Pop() interface{} {
	if s.IsEmpty() {
		return nil
	}
	index := len(s.items) - 1
	item := s.items[index]
	s.items = s.items[:index]
	return item
}

func (s *Stack) IsEmpty() bool {
	return len(s.items) == 0
}

func (s *Stack) Size() int {
	return len(s.items)
}
